validator-api/src/rewarding/mod.rs

```diff
@@ -149,132 +139,117 @@ impl Rewarder {
+                .get_average_mixnode_uptime_in_interval(
+                    rewarded_node.identity(),
+                    epoch.start_unix_timestamp(),
+                    epoch.end_unix_timestamp(),
+                )
+                .await?;
+
+            if uptime.u8() > 0 {
+                eligible_nodes.push(MixnodeToReward {
+                    identity: rewarded_node.mix_node.identity_key,
+                    total_delegations,
+                    params: NodeRewardParams::new(
+                        period_reward_pool,
+                        k.into(),
+                        0,
+                        circulating_supply,
+                        uptime.u8().into(),
+                        sybil_resistance_percent,
+                    ),
+                })
             }
-        } else {
-            info!("Tokenomics feature is OFF");
         }
 
         Ok(eligible_nodes)
     }
 
-    /// Obtains the lists of all mixnodes that were tested at least a single time
-    /// by the network monitor in the specified epoch.
+    /// Check whether every node, and their delegators, on the provided list were fully rewarded
+    /// in the specified interval.
     ///
-    /// # Arguments
+    /// It is used to deal with edge cases such that mixnode had exactly full page of delegations and
+    /// somebody created a new delegation thus causing the "last" delegator to possibly be pushed
+    /// onto the next page that the validator API was not aware of.
     ///
-    /// * `epoch`: the specified epoch.
-    async fn get_active_monitor_mixnodes(
+    /// * `eligible_mixnodes`: list of the nodes that were eligible to receive rewards.
+    /// * `rewarding_interval_nonce`: nonce associated with the current rewarding interval
+    async fn verify_rewarding_completion(
         &self,
-        epoch: Epoch,
-    ) -> Result<Vec<MixnodeStatusReport>, RewardingError> {
-        Ok(self
-            .storage
-            .get_all_active_mixnode_reports_in_interval(
-                epoch.start_unix_timestamp(),
-                epoch.end_unix_timestamp(),
-            )
-            .await?)
+        eligible_mixnodes: &[MixnodeToReward],
+        current_rewarding_nonce: u32,
+    ) -> (Vec<MixnodeToReward>, Vec<MixnodeToReward>) {
+        let mut unrewarded = Vec::new();
+        let mut further_delegators_present = Vec::new();
+        for mix in eligible_mixnodes {
+            let rewarding_status = self
+                .nymd_client
+                .get_rewarding_status(mix.identity.clone(), current_rewarding_nonce)
+                .await
+                .expect("error handling");
+            match rewarding_status.status {
+                // that case is super weird, it implies the node hasn't been rewarded at all!
+                // maybe the transaction timed out twice or something? In any case, we should attempt
+                // the reward for the final time!
+                None => unrewarded.push(mix.clone()),
```

Comment:
Double check this to figure out whether there's any legitimate case that this can happen in, or whether we should error out of this.